plugin/hooks.go

```diff
@@ -149,4 +154,41 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	// You should return the notification that you want to be sent to the user (normally the same notification).
+	//
+	// You can reject the notification returning a nil notification.
+	//
+	// You can modify the notification, returning a changed version of the notification.
+	//
+	// Minimum server version: 5.10
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server has replied with an error.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	//
+	// This is an internal API and it can change in the future.
+	//
+	// Minimum server version: 5.10
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
+
+	// PushNotificationAck is invoked when the server receive the confirmation
+	// of the reception of the push notification from the mobile device.
+	//
+	// You can return data that will be return to the mobile device as response to the ACK.
+	//
+	// Or you can return an error that will be returned to the mobile device as response to the ACK.
```

Comment:
What happens when the mobile device receives an error?

Reply:
Probably it will retry after X time, or so. I'm not sure, it depends on the mobile implementation. And probably if you return data, can be a modified version of the notification to update it in the device.

Reply:
I need to talk with @enahum to be sure what we expect here. Probably we will merge it when the whole thing is ready (the mobile apps code too).

Reply:
I'm with @hanzei here, I don't think we should care if this errors out other than logging the error. and the plugin should retry storing the ack if it fails for some reason.